### PR TITLE
feature/phone number format e164

### PR DESCRIPTION
**Pull request by dan-netizen**
Add the option to format the phone number to E164 standard using the google libphonenumber library, which should be packaged as a dependency in the final jar for the authenticator to work.
Add the option to add the phone number to the 'mobile_number' user attribute on successful phone number validation action.
Based on authenticator configuration, if the phone does not pass the required formatting checks, an error is shown and the user is asked to re-enter a valid phone number.

### CODE REVIEW

**Comment by svenseeberg** [comment]
> I could not find a way around making a jar with dependencies, which increases its size quite a bit.

The biggest issue here is that `io.quarkus` would be included, right? We only need this for development. If I understand https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#dependency-scope correctly, choosing the correct scope should mitigate the problem.

File: sms-authenticator/pom.xml [hunk 1]
```diff
@@ -18,6 +18,11 @@
 			<artifactId>keycloak-quarkus-server-app</artifactId>
 			<scope>provided</scope>
 		</dependency>
+		<dependency>
+			<groupId>com.googlecode.libphonenumber</groupId>
+			<artifactId>libphonenumber</artifactId>
+			<version>8.13.40</version>
```
Comment (svenseeberg) [inline sms-authenticator/pom.xml:24]: We should probably activate Github Dependabot to automatically bump versions and receive security updates.

File: sms-authenticator/src/main/resources/theme-resources/messages/messages_de.properties [hunk 2]
```diff
@@ -15,4 +15,8 @@ smsPhoneNumberInstructions=Bitte gib eine Telefonnummer inklusive L\u00E4ndervor
 smsPhoneUpdate=Empfange einen Einmal-Code per SMS auf dein Telefon.
 smsAuthenticator=SMS Authentifizierung
 
+numberFormatFailedToParse=Die Telefonnummer konnte nicht geparst werden. Bitte geben Sie es erneut ein.
```
Comment (svenseeberg) [inline sms-authenticator/src/main/resources/theme-resources/messages/messages_de.properties:18]:
```suggestion
numberFormatFailedToParse=Die Telefonnummer konnte nicht verarbeitet werden. Bitte versuchen Sie es erneut.
```

File: sms-authenticator/src/main/resources/theme-resources/messages/messages_de.properties [hunk 3]
```diff
@@ -15,4 +15,8 @@ smsPhoneNumberInstructions=Bitte gib eine Telefonnummer inklusive L\u00E4ndervor
...
+numberFormatNoMatchingFilters=Die Telefonnummer entspricht nicht den Filtern f�r den Telefonnummern Typ.
+numberFormatNumberInvalid=Ung�ltige Telefonnummer, bitte geben Sie sie erneut ein.
```
Comment (svenseeberg) [inline sms-authenticator/src/main/resources/theme-resources/messages/messages_de.properties:20]:
```suggestion
numberFormatFailedToParse=Die Telefonnummer konnte nicht verarbeitet werden. Bitte versuche es erneut.
numberFormatNoMatchingFilters=Diese Telefonnummer ist nicht gültig. Bitte nutze eine reguläre Mobilfunk- oder Festnetznummer.
numberFormatNumberInvalid=Ungültige Telefonnummer. Bitte versuche es erneut.
```

File: sms-authenticator/pom.xml (hunk 1 above)
Comment (svenseeberg) [inline sms-authenticator/pom.xml:24]: Just a general note: we should probably activate Github Dependabot to automatically bump versions and receive security updates.

File: sms-authenticator/src/main/resources/theme-resources/messages/messages_de.properties (hunk 3 above)
Comment (alexanderhofstaetter) [reply sms-authenticator/src/main/resources/theme-resources/messages/messages_de.properties:20]: Better german translations are needed

**Comment by dan-netizen** [comment]
I think I applied most of the suggestions, and replaced maven-assembly-plugin with maven-shade-plugin, that seemed to work better and the jar is now ~397KB down from ~13MB+.
Also added some other misc fixes and improvements.

**Comment by alexanderhofstaetter** [comment]
There seems to be some problems regarding encoding on your side:

image

File: .github/dependabot.yml [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+      interval: weekly
+    labels:
+      - "version-upgrade"
+    ignore:
+      - dependency-name: org.keycloak:keycloak-quarkus-server-app
+      - dependency-name: io.quarkus:quarkus-vertx-http-dev-ui-resources
+      - dependency-name: org.apache.maven.plugins:maven-surefire-plugin
+      - dependency-name: org.apache.maven.plugins:maven-compiler-plugin
+      - dependency-name: io.quarkus:quarkus-maven-plugin
+      - dependency-name: org.apache.maven.plugins:maven-assembly-plugin
```
Comment (melegiul) [inline .github/dependabot.yml:15]: Maven assembly was replaced. I think in general it would make more sense to update the maven plugins and keycloak dependencies project wide. But that is out of scope for this PR.

So please just replace ignore by access `com.googlecode.libphonenumber:libphonenumber`.

**Review by melegiul — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution. Finally the long missing proper input validation. :heart_eyes: 

Just some minor issues, see below. Then this should be ready to merge.

File: sms-authenticator/src/main/java/netzbegruenung/keycloak/authenticator/SmsAuthenticatorFactory.java [hunk 5]
```diff
@@ -86,7 +86,12 @@ public List<ProviderConfigProperty> getConfigProperties() {
...
-			new ProviderConfigProperty("hideResponsePayload", "Redacted API response log message", "Don't log API response body of SMS send request.", ProviderConfigProperty.BOOLEAN_TYPE, false)
+			new ProviderConfigProperty("hideResponsePayload", "Redacted API response log message", "Don't log API response body of SMS send request.", ProviderConfigProperty.BOOLEAN_TYPE, false),
+			new ProviderConfigProperty("mobileInputFieldPlaceholder", "Phone number input field placeholder", "The placeholder string user in the phone number input field", ProviderConfigProperty.STRING_TYPE, ""),
+			new ProviderConfigProperty("storeInAttribute", "Set phone number as attribute", "Sets the phone number as a user attribute.", ProviderConfigProperty.BOOLEAN_TYPE, false),
+			new ProviderConfigProperty("normalizePhoneNumber", "Format phone number", "Normalize the phone number using the E164 standard.", ProviderConfigProperty.BOOLEAN_TYPE, false),
+			new ProviderConfigProperty("numberTypeFilters", "Valid number type filters", "A comma separated list of valid number types.", ProviderConfigProperty.STRING_TYPE, ""),
```
[the remark this answers is not in the document]
Comment (alexanderhofstaetter) [reply sms-authenticator/src/main/java/netzbegruenung/keycloak/authenticator/SmsAuthenticatorFactory.java:93]: Also we should put a link to all the possible values in the help icon text, so user known what to input here.

File: sms-authenticator/src/main/resources/theme-resources/messages/messages_de.properties (hunk 3 above)
Comment (melegiul) [reply sms-authenticator/src/main/resources/theme-resources/messages/messages_de.properties]: Was addressed by fedcca05b59ba762cd5753df9aeb70f9d9efb188

File: sms-authenticator/pom.xml (hunk 1 above)
Comment (melegiul) [reply sms-authenticator/pom.xml:28]: Solved by 6922721b828f83d23627cbe16993c12a874b906f

File: sms-authenticator/pom.xml [hunk 6]
```diff
@@ -12,18 +12,25 @@
...
 		</dependency>
 		<!-- Necessary for proper execution of IDELauncher -->
 		<!-- Can be removed as part of the https://github.com/keycloak/keycloak/issues/22455 enhancement -->
 		<dependency>
 			<groupId>io.quarkus</groupId>
 			<artifactId>quarkus-vertx-http-dev-ui-resources</artifactId>
-			<scope>provided</scope>
```
Comment (melegiul) [inline sms-authenticator/pom.xml:26]: Please leave the scopes in sms-auth module.

In my original request I meant to remove scopes only from dependency management, which sits in the parent pom in the project root folder.

In summary:
- pom.xml -> no scopes
- sms-authenticator/pom.xml -> scopes

**Review by melegiul — CHANGES_REQUESTED** [review]
Thank you very much for your changes. Just one thing left to fix. I guess just a misunderstanding. See below.

**Comment by alexanderhofstaetter** [comment]
@melegiul can you provide a new release with all the changes?

**Comment by melegiul** [comment]
Yes, sure. I will create a new release not later than tomorrow. I just want to add another feature, but unrelated to SMS.

**Comment by melegiul** [comment]
And there it is :smiley: 
https://github.com/netzbegruenung/keycloak-mfa-plugins/releases/tag/v25.0.2
